cpp/src/arrow/util/async_iterator.h

```diff
@@ -0,0 +1,295 @@
+};
+
+template <typename T>
+static std::function<void(const Result<T>&)> MakeCallback(
+    std::shared_ptr<bool> finished) {
+  return [finished](const Result<T>& next_result) {
+    if (!next_result.ok()) {
+      *finished = true;
+    } else {
+      auto next = *next_result;
+      *finished = (next == IterationTraits<T>::End());
+    }
+  };
+}
+
+template <typename T>
+AsyncGenerator<T> AddReadahead(AsyncGenerator<T> source_generator, int max_readahead) {
+  // Using a shared_ptr instead of a lambda capture here because it's possible that
+  // the inner mark_finished_if_done outlives the outer lambda
+  auto finished = std::make_shared<bool>(false);
+  auto mark_finished_if_done = [finished](const Result<T>& next_result) {
+    if (!next_result.ok()) {
+      *finished = true;
+    } else {
+      auto next = *next_result;
+      *finished = (next == IterationTraits<T>::End());
+    }
+  };
+
+  std::queue<Future<T>> readahead_queue;
+  return [=]() mutable -> Future<T> {
+    if (readahead_queue.empty()) {
+      // This is the first request, let's pump the underlying queue
+      for (int i = 0; i < max_readahead; i++) {
+        auto next = source_generator();
+        next.AddCallback(mark_finished_if_done);
+        readahead_queue.push(std::move(next));
+      }
+    }
+    // Pop one and add one
+    auto result = readahead_queue.front();
+    readahead_queue.pop();
+    if (*finished) {
+      readahead_queue.push(Future<T>::MakeFinished(IterationTraits<T>::End()));
+    } else {
+      auto back_of_queue = source_generator();
+      back_of_queue.AddCallback(mark_finished_if_done);
+      readahead_queue.push(std::move(back_of_queue));
+    }
+    return result;
+  };
+}
+
+/// \brief Transforms an async generator using a transformer function returning a new
+/// AsyncGenerator
+///
+/// The transform function here behaves exactly the same as the transform function in
+/// MakeTransformedIterator and you can safely use the same transform function to
+/// transform both synchronous and asynchronous streams.
+template <typename T, typename V>
+AsyncGenerator<V> TransformAsyncGenerator(AsyncGenerator<T> generator,
+                                          Transformer<T, V> transformer) {
+  return TransformingGenerator<T, V>(generator, transformer);
+}
+
+namespace detail {
+
+template <typename T>
+struct BackgroundIteratorPromise : ReadaheadPromise {
+  ~BackgroundIteratorPromise() override {}
+
+  explicit BackgroundIteratorPromise(Iterator<T>* it) : it_(it) {}
+
+  bool Call() override {
+    auto next = it_->Next();
+    auto finished = next == IterationTraits<T>::End();
+    out_.MarkFinished(std::move(next));
+    return finished;
+  }
+
+  void End() override { out_.MarkFinished(IterationTraits<T>::End()); }
+
+  Iterator<T>* it_;
+  Future<T> out_ = Future<T>::Make();
+};
+
+}  // namespace detail
+
+/// \brief Async generator that iterates on an underlying iterator in a
+/// separate thread.
```

Comment:
No, not in an executor at the moment, and probably never will.  At some point, I think the filesystem interfaces will support async and then the filesystems themselves can assign executors where appropriate.  Once that is done there will be no need for this "backgrounding" class.  The background threads certainly don't belong on the CPU thread pool.  We could create a "background" thread pool but it would have to be large enough to support the maximum parallelism we want from the filesystem (which may be greater than the # of CPUs in a slow networked I/O case).

This generator (I will change that name per next comment) can indeed be called from several threads at once.  Although there is no need for it and we can do away with that requirement if it helps us save on concurrency cost.  I'll write a separate note on the parallelism of these generators.  The underlying iterator will never be iterated on by any more than the background thread.